cvmfs/cache_ram.cc

```diff
@@ -0,0 +1,250 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+#include "cvmfs_config.h"
+#include "cache_ram.h"
+
+#include <errno.h>
+#include <assert.h>
+#include <new>
+#include <string.h>
+
+#include "kvstore.h"
+#include "smalloc.h"
+#include "util/posix.h"
+#include "util_concurrency.h"
+
+using namespace std;  // NOLINT
+
+namespace cache {
+
+
+int RamCacheManager::AddFd(const ReadOnlyFd &fd) {
```

Comment:
The above would make assignment of a FD to be constant time (well, plus a lock...) with the tradeoff of extra memory usage.

To bound the memory usage, one could bound the maximum size of the `free_fd_` list.  This would cause a sequential scan to refill it if it is empty but `open_fds_` has known vacancies.  You'd also need to track the occupancy of `open_fds_` - but that can also be done in constant time with simple counters.

At that point, it would be convenient to split it out to a separate "fd tracker" object, as I'm facing the exact same issues with my `librados`-based cache.  Once we can uplift the CVMFS code base to C++11, such an object would be a prime candidate for using lock-free, atomics-based methods.